File: SimpleSchedulerApiModels/Schedule.cs
namespace SimpleSchedulerApiModels;

public class Schedule
{
    public Schedule() { }

    public Schedule(long id, bool isActive, long workerID,
        bool sunday, bool monday, bool tuesday, bool wednesday, bool thursday, bool friday, bool saturday,
        TimeSpan? timeOfDayUTC, TimeSpan? recurTime, TimeSpan? recurBetweenStartUTC, TimeSpan? recurBetweenEndUTC,
        bool oneTime
    )
    {
        ID = id;
        IsActive = isActive;
        WorkerID = workerID;
        Sunday = sunday;
        Monday = monday;
        Tuesday = tuesday;
        Wednesday = wednesday;
        Thursday = thursday;
        Friday = friday;
        Saturday = saturday;
        TimeOfDayUTC = timeOfDayUTC;
        RecurTime = recurTime;
        RecurBetweenStartUTC = recurBetweenStartUTC;
        RecurBetweenEndUTC = recurBetweenEndUTC;
        OneTime = oneTime;
    }

    public long ID { get; set; }
    public bool IsActive { get; set; }
    public long WorkerID { get; set; }
    public bool Sunday { get; set; }
    public bool Monday { get; set; }
    public bool Tuesday { get; set; }
    public bool Wednesday { get; set; }
    public bool Thursday { get; set; }
    public bool Friday { get; set; }
    public bool Saturday { get; set; }
    public TimeSpan? TimeOfDayUTC { get; set; }
    public TimeSpan? RecurTime { get; set; }
    public TimeSpan? RecurBetweenStartUTC { get; set; }
    public TimeSpan? RecurBetweenEndUTC { get; set; }
    public bool OneTime { get; set; }

    public string GetFormatted()
    {
        string days;

        if (Sunday && Monday && Tuesday && Wednesday && Thursday && Friday && Saturday)
        {
            days = "Every day";
        }
        else if (!Sunday && Monday && Tuesday && Wednesday && Thursday && Friday && !Saturday)
        {
            days = "Weekdays";
        }
        else if (Sunday && !Monday && !Tuesday && !Wednesday && !Thursday && !Friday && Saturday)
        {
            days = "Weekends";
        }
        else
        {
            days = string.Join(' ', new[] {
                Sunday ? "Su" : "__",
                Monday ? "Mo" : "__",
                Tuesday ? "Tu" : "__",
                Wednesday ? "We" : "__",
                Thursday ? "Th" : "__",
                Friday ? "Fr" : "__",
                Saturday ? "Sa" : "__"
            });
        }

        string times = "Unknown";
        const string TS_FORMAT = @"hh\:mm";
        if (TimeOfDayUTC.HasValue)
        {
            times = $"at { TimeOfDayUTC.Value.ToString(TS_FORMAT) }";
        }
        else if (RecurTime.HasValue)
        {
            times = GetFormattedTimeSpan(RecurTime.Value);
        }

        if (RecurBetweenStartUTC.HasValue && RecurBetweenEndUTC.HasValue)
        {
            times += $" between { RecurBetweenStartUTC.Value.ToString(TS_FORMAT) } and { RecurBetweenEndUTC.Value.ToString(TS_FORMAT) }";
        }
        else if (RecurBetweenStartUTC.HasValue)
        {
            times += $" starting at { RecurBetweenStartUTC.Value.ToString(TS_FORMAT) }";
        }
        else if (RecurBetweenEndUTC.HasValue)
        {
            times += $" until { RecurBetweenEndUTC.Value.ToString(TS_FORMAT) }";
        }

        return $"{ days } [{ times }]";
    }

    public static Schedule GetDummySchedule(long workerID)
    {
        return new(
            id: 0,
            isActive: false,
            workerID: workerID,
            sunday: true,
            monday: true,
            tuesday: true,
            wednesday: true,
            thursday: true,
            friday: true,
            saturday: true,
            timeOfDayUTC: null,
            recurTime: TimeSpan.FromHours(1),
            recurBetweenStartUTC: null,
            recurBetweenEndUTC: null,
            oneTime: false
        );
    }

    private static string GetFormattedTimeSpan(TimeSpan ts)
    {
        if (ts.Hours == 1 && ts.Minutes == 0) { return "every hour"; }
        if (ts.Hours > 1 && ts.Minutes == 0) { return $"every { ts.Hours } hours"; }
        if (ts.Hours > 0 && ts.Minutes > 0) { return $"every {ts:hh\\:mm}"; }
        if (ts.Minutes == 1) { return "every minute"; }
        if (ts.Minutes > 1) { return $"every { ts.Minutes} minutes"; }
        return "unknown";
    }
}

File: SimpleSchedulerCommandLine/models/schedule.go
package models

import (
	"fmt"
	"strconv"
	"strings"
)

type Schedule struct {
	ID                   int64  `json:"id"`
	IsActive             bool   `json:"isActive"`
	WorkerID             int64  `json:"workerID"`
	Sunday               bool   `json:"sunday"`
	Monday               bool   `json:"monday"`
	Tuesday              bool   `json:"tuesday"`
	Wednesday            bool   `json:"wednesday"`
	Thursday             bool   `json:"thursday"`
	Friday               bool   `json:"friday"`
	Saturday             bool   `json:"saturday"`
	TimeOfDayUTC         string `json:"timeOfDayUTC"`
	RecurTime            string `json:"recurTime"`
	RecurBetweenStartUTC string `json:"recurBetweenStartUTC"`
	RecurBetweenEndUTC   string `json:"recurBetweenEndUTC"`
	OneTime              bool   `json:"oneTime"`
}

func tern(b bool, first string, second string) string {
	if b {
		return first
	}
	return second
}

func getFormattedTime(ts string) string {
	if len(ts) == 0 {
		return ""
	}

	fields := strings.Split(ts, ":")
	hours, _ := strconv.Atoi(fields[0])
	minutes, _ := strconv.Atoi(fields[1])

	if hours == 1 && minutes == 0 {
		return "every hour"
	}
	if hours > 1 && minutes == 0 {
		return fmt.Sprintf("every %v hours", hours)
	}
	if hours > 0 && minutes > 0 {
		return fmt.Sprintf("every %s", ts)
	}
	if minutes == 1 {
		return "every minute"
	}
	if minutes > 1 {
		return fmt.Sprintf("every %v minutes", minutes)
	}
	return "unknown"
}

func (sch *Schedule) GetFormatted() string {
	var days string
	if sch.Sunday && sch.Monday && sch.Tuesday && sch.Wednesday && sch.Thursday && sch.Friday && sch.Saturday {
		days = "Every day"
	} else if !sch.Sunday && sch.Monday && sch.Tuesday && sch.Wednesday && sch.Thursday && sch.Friday && !sch.Saturday {
		days = "Weekdays"
	} else if sch.Sunday && !sch.Monday && !sch.Tuesday && !sch.Wednesday && !sch.Thursday && !sch.Friday && sch.Saturday {
		days = "Weekends"
	} else {
		days = fmt.Sprintf("%s %s %s %s %s %s %s",
			tern(sch.Sunday, "Su", "__"),
			tern(sch.Monday, "Mo", "__"),
			tern(sch.Tuesday, "Tu", "__"),
			tern(sch.Wednesday, "We", "__"),
			tern(sch.Thursday, "Th", "__"),
			tern(sch.Friday, "Fr", "__"),
			tern(sch.Saturday, "Sa", "__"),
		)
	}

	times := "Unknown"

	if len(sch.TimeOfDayUTC) > 0 {
		times = fmt.Sprintf("at %s", sch.TimeOfDayUTC)
	} else if len(sch.RecurTime) > 0 {
		times = getFormattedTime(sch.RecurTime)
	}

	if len(sch.RecurBetweenStartUTC) > 0 && len(sch.RecurBetweenEndUTC) > 0 {
		times += fmt.Sprintf(" between %s and %s", sch.RecurBetweenStartUTC, sch.RecurBetweenEndUTC)
	} else if len(sch.RecurBetweenStartUTC) > 0 {
		times += fmt.Sprintf(" starting at %s", sch.RecurBetweenStartUTC)
	} else if len(sch.RecurBetweenEndUTC) > 0 {
		times += fmt.Sprintf(" until %s", sch.RecurBetweenEndUTC)
	}

	return fmt.Sprintf("%s [%s]", days, times)
}
